### DISCUSSION TITLE
Add a prompt for unsaved changes to `beforeunload`

### DISCUSSION THREAD

**Discussion by izucken** [directus/directus#22677] [Feature Requests]
### Summary

I tried looking through the issues and discussions for this. Closest case I could find is https://github.com/directus/directus/discussions/11666, it proposes another solution in a somewhat different context.

---

I had a content editor contact me about some changes they apparently lost. In their scenario, they made some changes or added a linked entity (the one for which a side panel opens), but it was not entirely obvious to them that such action may necessitate saving the root entity as well. In our case, forms are not too big, but related entity editing can make current application state unclear to users (after all it seems like you clicked a checkmark button called "save").

When you try to navigate away from the form within the app, directus UI will helpfully ask you to stay or discard the changes. If you accidentally navigate away from directus within the tab, you can go back and recover form state, which is also useful and seems to work for the entire tab history (although it can behave not entirely obviously if e.g. user navigates back manually rather than via history api).

But, if you close this tab entirely, state will be lost. I thought maybe a simple browser prompt can help alleviate this. Of course you can avoid this with good discipline, saving often or changing stuff more granularly, but I am talking about accidents here.

If you think this is reasonable I am open to implementing it, feels fairly straightforward.

### Motivation

Help user to avoid accidentally losing their work.

### Detailed Design

Prevent closing a tab with an unsaved form using a prompt bound to a `beforeunload` event. Detect unsaved forms using the same logic used for prompts on internal navigation.

### Requirements List

Not sure.

### Drawbacks

Such prompts may be disabled by the user, but this is clearly a case of manual opt out. If the prompt is often discarded, iirc the browser may not show them again for the domain, which also seems like an opt out.

### Alternatives

I believe this could be solved by some sort of a snapshot for the form in, say, local storage, but is likely to be fairly complicated, considering multiple open tabs, history states or user sessions.

Another possible alternative is a UX change to make unsaved form more clear and somehow disambiguate intermediate saves (in subforms/side panels) from the overall form (root) save. The state could made be more obvious with screen decoration, say a signaling color line on the side of the form (I remember doing something like that in the past but for other reasons), not sure about disambiguation.

### Adoption Strategy

I think this is supposed to be purely a UI change.

**Comment by izucken** [comment]
Forgot to mention and thank you Rijk and others for a great piece of software... The first time ever I am not befuddled by a CMS-like system...

**Comment by hanneskuettner** [comment]
Having a look through the issues it looks like this comes very close to your request, since it would effectively be implemented the same https://github.com/directus/directus/issues/22264

**Reply by izucken** [reply]
Thanks for finding it, it really is directly related.
